File: backend/schemas/subject.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class SubjectCreate(BaseModel):
    name: str


class SubjectUpdate(BaseModel):
    name: str


class SubjectResponse(BaseModel):
    id: int
    teacher_id: int
    name: str
    course_code: str
    # How many students are enrolled, and how many of those already have a
    # complete (>= MIN_BASELINE_PAPERS) baseline - the roster progress bar.
    student_count: int
    baseline_ready_count: int
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class RosterStudent(BaseModel):
    id: int
    name: str
    email: str | None
    status: str
    baseline_paper_count: int
    baseline_ready: bool

    model_config = ConfigDict(from_attributes=True)


class RosterResponse(BaseModel):
    subject_id: int
    subject_name: str
    course_code: str
    students: list[RosterStudent]


class BatchUploadSkip(BaseModel):
    row: int
    reason: str


class BatchUploadResult(BaseModel):
    created_count: int
    skipped: list[BatchUploadSkip]

File: frontend/types/subject.ts
export interface Subject {
  id: number;
  teacher_id: number;
  name: string;
  /** Unique, server-generated. Goes in the Subject Code column of the batch CSV. */
  course_code: string;
  /** Active enrollments only. */
  student_count: number;
  /** Active enrollments whose student has a complete (3-paper) baseline. */
  baseline_ready_count: number;
  created_at: string;
}

export interface CreateSubjectRequest {
  name: string;
}

export interface UpdateSubjectRequest {
  name: string;
}

export type EnrollmentStatus = "active" | "inactive";

export interface RosterStudent {
  id: number;
  name: string;
  email: string | null;
  status: EnrollmentStatus;
  baseline_paper_count: number;
  baseline_ready: boolean;
}

export interface Roster {
  subject_id: number;
  subject_name: string;
  course_code: string;
  students: RosterStudent[];
}

export interface BatchUploadResult {
  created_count: number;
  /** `row` is 1-based and counts the header as row 1, like a spreadsheet. */
  skipped: { row: number; reason: string }[];
}
